add support for list network area and test (#381)

larrytamnjong opened the pull request.

larrytamnjong commented:
@marcin-krystianc

marcin-krystianc commented on Consul.Test/OperatorTest.cs:
```diff
@@ -104,5 +104,11 @@ public async Task Operator_CreateArea()
+        public async Task Operator_AreaList()
+        {
+            var req = await _client.Operator.AreaList();
+            Assert.NotNull(req.Response);
```
To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).

larrytamnjong replied:
Thanks, done ✅

marcin-krystianc replied:
Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.

marcin-krystianc reviewed (commented):
The test are failing now with:
```
Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
            Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
        }	        }
        [EnterpriseOnlyFact]
        public async Task Operator_AreaList()
        {
            await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });

            var req = await _client.Operator.AreaList();
            Assert.NotNull(req.Response);
Member
@[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).

Contributor
Author
@[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
Thanks, done ✅

Member
@[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.

@marcin-krystianc	Reply...
        }
    }	    }
}	}
```

Maybe try using a random name every time you create an area.

larrytamnjong commented:
> The test are failing now with:
> 
> ```
> Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
>             Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
>         }	        }
>         [EnterpriseOnlyFact]
>         public async Task Operator_AreaList()
>         {
>             await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });
> 
>             var req = await _client.Operator.AreaList();
>             Assert.NotNull(req.Response);
> Member
> @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
> To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).
> 
> Contributor
> Author
> @[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
> Thanks, done ✅
> 
> Member
> @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
> Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.
> 
> @marcin-krystianc	Reply...
>         }
>     }	    }
> }	}
> ```
> 
> Maybe try using a random name every time you create an area.

Thanks, @marcin-krystianc your right the error occurred because I tried adding an operator area with the name "dc2," which was already created during the "create operator area" test. I’ve fixed that issue now. Hopefully, it's okay now.

marcin-krystianc commented:
> > The test are failing now with:
> > ```
> > Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
> >             Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
> >         }	        }
> >         [EnterpriseOnlyFact]
> >         public async Task Operator_AreaList()
> >         {
> >             await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });
> > 
> >             var req = await _client.Operator.AreaList();
> >             Assert.NotNull(req.Response);
> > Member
> > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
> > To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).
> > 
> > Contributor
> > Author
> > @[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
> > Thanks, done ✅
> > 
> > Member
> > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
> > Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.
> > 
> > @marcin-krystianc	Reply...
> >         }
> >     }	    }
> > }	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe try using a random name every time you create an area.
> 
> Thanks, @marcin-krystianc your right the error occurred because I tried adding an operator area with the name "dc2," which was already created during the "create operator area" test. I’ve fixed that issue now. Hopefully, it's okay.

Using fixed name is not ideal, I guess it is impossible to run that test twice in a row? The test should be implemented in a way, that makes it possible to run it multiple times without any issue.

larrytamnjong commented:
> > > The test are failing now with:
> > > ```
> > > Consul.Test.OperatorTest.Operator_CreateArea: Consul.ConsulRequestException : Unexpected response, status code InternalServerError: area "d0c9af6b-363b-69b3-a1a9-50089871df13" with peer datacenter "dc2" already exists
> > >             Assert.NotNull(response.Response);	            Assert.NotNull(response.Response);
> > >         }	        }
> > >         [EnterpriseOnlyFact]
> > >         public async Task Operator_AreaList()
> > >         {
> > >             await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = "dc2", UseTLS = false, RetryJoin = null });
> > > 
> > >             var req = await _client.Operator.AreaList();
> > >             Assert.NotNull(req.Response);
> > > Member
> > > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc 3 hours ago
> > > To make this test more exhaustive, it would be nice to create an area first and then assert that the query result is non-empty (instead of not null).
> > > 
> > > Contributor
> > > Author
> > > @[larrytamnjong](https://github.com/larrytamnjong) larrytamnjong 29 minutes ago
> > > Thanks, done ✅
> > > 
> > > Member
> > > @[marcin-krystianc](https://github.com/marcin-krystianc) marcin-krystianc now
> > > Pleas also updated the test condition, from not null to non-empty. Please inspect also a result fields if they are filled in.
> > > 
> > > @marcin-krystianc	Reply...
> > >         }
> > >     }	    }
> > > }	}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Maybe try using a random name every time you create an area.
> > 
> > 
> > Thanks, @marcin-krystianc your right the error occurred because I tried adding an operator area with the name "dc2," which was already created during the "create operator area" test. I’ve fixed that issue now. Hopefully, it's okay.
> 
> Using fixed name is not ideal, I guess it is impossible to run that test twice in a row? The test should be implemented in a way, that makes it possible to run it multiple times without any issue.

Thanks, @marcin-krystianc, and apologies for the back-and-forth. I’ve added a method to generate a random dc name for each test run. The tests are still in queue. Let me know if this approach works or if there's anything else you'd like me to adjust.

marcin-krystianc commented on Consul.Test/OperatorTest.cs:
```diff
@@ -17,13 +17,21 @@
+        internal static string GeneratePeerDatacenterName()
+        {
+            int randomNumber = Random.Next(100, 1000);
+            return $"ran-dc-{randomNumber}";
```
Single random number might be not enough to prevent from conflicts/duplicates. We've already a method `KVTest.GenerateTestKeyName()` which is being used in many places in tests already. It generates 16 character long random string. Please use it instead.

larrytamnjong replied:
@marcin-krystianc thanks, you can take another look now.

marcin-krystianc commented on Consul.Test/OperatorTest.cs:
```diff
@@ -17,13 +17,15 @@
 {
     public class OperatorTest : BaseFixture
     {
+
```
nit: extra empty line

marcin-krystianc commented on Consul.Test/OperatorTest.cs:
```diff
@@ -100,9 +102,20 @@ public async Task Segment_List()
+            await _client.Operator.CreateArea(new AreaRequest { PeerDatacenter = peerDataCenter, UseTLS = false, RetryJoin = null });
+
+            var req = await _client.Operator.AreaList();
+            Assert.NotEmpty(req.Response);
```
Just one more thing, can we test if the fields in the response are not null? 
So far we know that we can create and list areas, but it would be great if we assert that the values returned by the list operation correspond to the values used in the create operation.

larrytamnjong replied:
✅ done

marcin-krystianc replied:
Can you also try filling the `RetryJoin` field and then validating the content of it when returned?

marcin-krystianc commented on Consul/Interfaces/IOperatorEndpoint.cs:
```diff
@@ -47,6 +47,7 @@ public interface IOperatorEndpoint
         Task<WriteResult<string>> CreateArea(AreaRequest area, WriteOptions q, CancellationToken ct = default);
         Task<WriteResult<string>> CreateArea(AreaRequest area, CancellationToken ct = default);
-
+        Task<QueryResult<List<Area>>> AreaList(CancellationToken cancellationToken = default);
```
We should keep the method names consistent with others. I think `AreaList` is fine, but we should rename `CreateArea` to `AreaCreate`. It would help with `Area` method discoverability (user just types Area and presses autocomplete to see available options). What do you think?

larrytamnjong replied:
Makes a lot to sense ✅

marcin-krystianc commented on Consul.Test/OperatorTest.cs:
```diff
@@ -98,11 +101,26 @@ public async Task Segment_List()
+
+            var req = await _client.Operator.AreaList();
+            Assert.NotEmpty(req.Response);
+            Assert.Equal(area.PeerDatacenter, req.Response.First().PeerDatacenter);
```
There can be multiple items in the result list, so the first one is not necessarily the one we are looking for. You need to look for the right one, e.g.: 
```
var result = req.Response.Single(x => x.PeerDatacenter == area.peerDataCenter);
Assert.Equal(area.UseTLS, result.UseTLS);
Assert.Equal(area.RetryJoin, result.RetryJoin);  
```

larrytamnjong replied:
@marcin-krystianc I'm not sure if testing RetryJoin is sufficient with this approach ```Assert.Equal(area.RetryJoin, result.RetryJoin);``` 
or if it would be better to test each element individually in the array, like this: ```for (int i = 0; i < retryJoinAddresses.Length; i++)
    {
        Assert.Equal(retryJoinAddresses[i], createdArea.RetryJoin[i]);
    }```

marcin-krystianc replied:
> @marcin-krystianc I'm not sure if testing RetryJoin is sufficient with this approach `Assert.Equal(area.RetryJoin, result.RetryJoin);` or if it would be better to test each element individually in the array, like this: `for (int i = 0; i < retryJoinAddresses.Length; i++) { Assert.Equal(retryJoinAddresses[i], createdArea.RetryJoin[i]); }`

There is an overload for `IEnumerable`:
```
    public static void Equal<T>(
      IEnumerable<T> expected,
      IEnumerable<T> actual,
      IEqualityComparer<T> comparer)
```

so I guess it will do the right thing. 
You can try to define manually two lists  and see how it reacts to different item order in both lists.

larrytamnjong replied:
![image](https://github.com/user-attachments/assets/b6f6dc4e-bd45-4277-b8c8-b829367d515f)
Tests 1 and 3 are passing, but test 2 is failing so its ok. I have pushed the changes please let me know if there's anything else to change.

marcin-krystianc reviewed (approved):
It's perfect now.
